Add expired client identifier field and checks to improve old clientId check perf

heliocliu opened the pull request:
#3675
Change the old/expired clientId check to use the quorum instead of by keeping a list of old clientIds.  This works by adding a shouldHaveLeft field to the ISequencedClient in the quorum, where a reconnecting client will set its old clientId shouldHaveLeft to true (if it still exists in the client).  Then when we process a message, instead of checking against a list of old clientIds, we check if the clientId exists in the quorum (messages should only be coming from clients in the quorum) and if the clientId has been marked should have left.  An inactive clientId in this situation would suggest a message from an old connection of the same client (which shouldn't happen).

heliocliu committed:
add isActive field and checks

heliocliu commented:
Will check in server/client parts separately; both included for now to show the entire change

tanviraumi commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -1453,6 +1460,25 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
+         // or we previously marked inactive
+         if (!local) {
+            const client = this._protocolHandler?.quorum.getMember(message.clientId);
+            const missingFromQuorum = client === undefined;
```
This should never happen. By definition, the client should always be in the quorum.

tanviraumi commented on server/routerlicious/packages/protocol-definitions/src/clients.ts:
```diff
@@ -26,10 +26,9 @@ export interface IClient {
     client: IClient;
-
     sequenceNumber: number;
+    isActive: boolean;
 }
```
I am a little hesitant to put a logging/telemetry specific required field to protocol-definition. Also a client inside the quorum is always active by definition. So adding a isActive field seems weird to me.

Because we are only trying to add log, I would suggest tracking it separately. Referring to #3675, this has some memory impact. But probably not as much as we think. Unless we have logging already showing 100 inactive clients, I am not so convinced that it will happen. Cause you can always clean up that list when a leave message shows up.

Now if we are still too worried about memory implication, an alternate is to extend the ISequencedClient interface in container and use it to track isActive. Then we get the best of both world.

heliocliu replied:
moved it all into container to avoid changing server

heliocliu committed:
move all changes into container away from server

heliocliu committed:
remove missing from quorum check

this check doesn't work for older messages. could probably add a way to check if it's recent enough, but that seems kind of iffy and isn't necessary for the case we're trying to cover with telemetry here

anthony-murphy commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -1453,6 +1465,21 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
+         if (!local) {
+            const client: IActiveSequencedClient | undefined =
+                this._protocolHandler?.quorum.getMember(message.clientId);
+            if (client?.isActive === false) {
```
this won't catach the issue. both clients were still active from the servers persepctive which is what quorum represents

anthony-murphy replied:
if a client id is not active from the server's perspective, i'm pretty sure the server will nack any ops for that client, so this should also never happen

heliocliu replied:
Isn't this what we're looking for though?  (Cases where these ops should have been nacked, but aren't)
The `isActive` flag is set locally on our side, so the only input the server has on it here is if the client gets removed from the quorum.  This is the same place we added telemetry before, except this time the quorum is being used to remove clientIds that should no longer matter instead of keeping them indefinitely until they are bumped out of the MRU.

anthony-murphy replied:
No. That's not what happens. The old client id has not had a leave message sent by the server, as it was the client who decided it needed to reconnect. We see the first set of ops from the old client id, then the resubmitted ops from the new client id, and only after that do we see a leave message for the old client id.

anthony-murphy replied:
i updated #3627 to make this explicit

heliocliu replied:
Here are the steps as I understand and how this PR fits into it, which I _think_ matches what's in #3627.  If you could help me with which part I'm missing, I'd appreciate it.

1. Old client join message
2. First set of ops sent from old client id
3. New client join message
3a. During new client connect, we hit the change in L1398 above.  When we change our client id, we see the old client is still in the quorum (because there has not been an old client leave message yet), so we mark it as `isActive = false`, but don't remove it from the quorum.  
4. Second set of ops sent from old client id
4a. New client sees second set of ops and hits the change here.  It doesn't recognize them as coming from the same client (`!local`), but because old client has not had its leave message yet (and is still in the quorum), new client will find it in the quorum marked `!isActive` (which was done earlier on new client connect).  It then reports telemetry.
5. Duplicate second set of ops are sent from new client (because the ops sent from old client were not recognized)
6. Old client leave message

vladsud replied:
Yep, that looks correct. :) There might be many reasons for the issue, but this check is equivalent to what an keeping at array of 100 clientIDs did - we should evaluate equivalence here only.
If there are more checks or more cases to track, we should do it in separate PR.

vladsud commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -1453,6 +1465,21 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
+         if (!local) {
+            const client: IActiveSequencedClient | undefined =
+                this._protocolHandler?.quorum.getMember(message.clientId);
+            if (client?.isActive === false) {
```
I do not think this check is correct.
I'd make it
if (client === undefined || client.isActive === false)
i.e. the client that sends ops should be in quorum!

That said, this would work only if we make this change for !isSystemMessage(message), i.e. exclude "join" & "leave" messages as they are odd - we need to give this.protocolHandler.processMessage a chance to process join for this check to apply to it, but leave is other way around.

vladsud replied:
BTW, I'd consider dropping if (!local) check for this - the invariant is still valid - the client that is sending ops should be in quorum, and can't be non-active. Local or not is secondary.

heliocliu replied:
How does the `client === undefined` check work with unrelated clients playing back older ops to catch up to the latest?  I had something similar in the first iteration and it was getting hit during the UTs and failing them, so I wasn't super sure if it wouldn't work outside of the live scenario, or if it was an issue with the local test -licious variant (which seemed unlikely)

vladsud replied:
Your current check only caches cases that are still in quorum, i.e. server still thinks that client is alive (otherwise server would generate leave op for it and it would be deleted from quorum, loosing active flag tracing).

Increasing check to include client === undefined would not help to catch client error (or rather - client-only errors), because client missing from quorum should not happen - server should not allow to accept and retransmit ops from clients that are not in quorum. But it might be still valuable (in terms of increasing coverage), as maybe there are server bugs we are against, or maybe re-ordering of ops happens somewhere else (in driver?) that may lead to this condition - hitting it would be useful indicator of where to keep digging.

You likely hit problems with join & leave messages, as join message would be coming from client that is still not in quorum - we add it to quorum down when calling into protocolHandler. Reordering check with protocolHandler would not help as it would fix it for join op, but make it fail for leave op. That's the reason I think the best check is something like that::

if (client === undefined && op.type !== MessageType.ClientJoin || client?.isActive === false) {

If it's not the case, then it would be interesting to learn why you hit problems. I'm not aware of any other reason why it's wrong check to have.

heliocliu replied:
problem was just that the check wasn't handling noops (because message.clientId was null 🙃). all good after handling that case

vladsud commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -1453,6 +1465,21 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
+            const client: IActiveSequencedClient | undefined =
+                this._protocolHandler?.quorum.getMember(message.clientId);
+            if (client?.isActive === false) {
+                this.logger.sendErrorEvent({
```
I prefer fail-fast. I.e. critical error on container. The reason - container is screwed after that, it's much better to fail it (forever) but allow user to recover content via copy / paste. It's worse if we allow things to continue. This case might be not that bad, but in other cases we did not even had eventual consistency (like creating data store with same ID) - two clients continued, but refreshing one would result in loss of content. While it does not feel like we have that kind of issue here, I find it better to fail safety then take chances.

anthony-murphy replied:
just started a thread with matt and helio about this. i agree we should close the container. is some cases this might even prevent corruption, as it will block resubmit. it's still possible the old ops are incorrect and corrupt the document anyway

heliocliu committed:
Update container.ts

vladsud commented:
Looks great!

heliocliu committed:
fix error to contain expected props

heliocliu committed:
Merge branch 'main' into ImproveClientIdLogging

heliocliu committed:
handle clientId:null case for noops
